apps/prairielearn/src/lib/course.ts

```diff
@@ -0,0 +1,231 @@
+  return await sqldb.queryRows(sql.select_owners, { course_id }, UserSchema);
+}
+
+export function getLockNameForCoursePath(coursePath: string): string {
+  return `coursedir:${coursePath}`;
+}
+
+export async function selectCourseById(course_id: string): Promise<Course> {
+  return await sqldb.queryRow(sql.get_course_data, { course_id }, CourseSchema);
+}
+
+export async function getCommitHash(coursePath: string): Promise<string> {
+  try {
+    const { stdout } = await promisify(exec)('git rev-parse HEAD', {
+      cwd: coursePath,
+      env: process.env,
+    });
+    return stdout.trim();
+  } catch (err) {
+    throw error.makeWithData(`Could not get git status; exited with code ${err.code}`, {
+      stdout: err.stdout,
+      stderr: err.stderr,
+    });
+  }
+}
+
+/**
+ * Loads the current commit hash from disk and stores it in the database. This
+ * will also add the `commit_hash` property to the given course object.
+ */
+export async function updateCourseCommitHash(course: {
+  id: string;
+  path: string;
+}): Promise<string> {
+  const hash = await getCommitHash(course.path);
+  await sqldb.queryAsync(sql.update_course_commit_hash, {
+    course_id: course.id,
+    commit_hash: hash,
+  });
+  return hash;
+}
+
+/**
+ * If the provided course object contains a commit hash, that will be used;
+ * otherwise, the commit hash will be loaded from disk and stored in the
+ * database.
+ *
+ * This should only ever really need to happen at max once per course - in the
+ * future, the commit hash will already be in the course object and will be
+ * updated during course sync.
+ */
+export async function getOrUpdateCourseCommitHash(course: {
+  id: string;
+  path: string;
+  commit_hash?: string | null;
+}): Promise<string> {
+  return course.commit_hash ?? (await updateCourseCommitHash(course));
+}
+
+export async function pullAndUpdate({
+  courseId,
+  userId,
+  authnUserId,
+  path,
+  branch,
+  repository,
+  commit_hash,
+}: {
+  courseId: string;
+  userId: string;
+  authnUserId: string;
+  path?: string | null;
+  branch?: string | null;
+  repository?: string | null;
+  commit_hash?: string | null;
+}): Promise<{ jobSequenceId: string; jobPromise: Promise<void> }> {
+  const serverJob = await createServerJob({
+    courseId,
+    userId,
+    authnUserId,
+    type: 'sync',
+    description: 'Pull from remote git repository',
+  });
+
+  const gitEnv = process.env;
+  if (config.gitSshCommand != null) {
+    gitEnv.GIT_SSH_COMMAND = config.gitSshCommand;
+  }
+
+  const jobPromise = serverJob.execute(async (job) => {
+    if (path === undefined || branch === undefined || repository === undefined) {
+      const course_data = await selectCourseById(courseId);
+      path = course_data.path;
+      branch = course_data.branch;
+      repository = course_data.repository;
+      commit_hash = course_data.commit_hash;
+    }
+    if (!path) {
+      job.fail('Path is not set for this course. Exiting...');
+      return;
```

Comment:
Aside: I was poking into why this `return` was required; I thought it shouldn't have been since `ServerJob#fail()` is marked as returning `never`. However, it looks like this is a design limitation of TypeScript: https://github.com/microsoft/TypeScript/issues/50363

If you declare `job: ServerJob` up on line 133, we could get rid of some returns, but that's not worth the effort! Just thought this was an interesting factoid about TypeScript.